File: backend/app/models/project.py
"""
Pydantic Models — Project

Hierarchy: Organization → Project → Worksites
"""

from pydantic import BaseModel, Field
from typing import Optional
from datetime import date, datetime
from decimal import Decimal
from enum import Enum


class ProjectStatus(str, Enum):
    DRAFT = "draft"
    PLANNING = "planning"
    ACTIVE = "active"
    REVIEW = "review"
    COMPLETE = "complete"
    ON_HOLD = "on_hold"
    CANCELLED = "cancelled"


# ── Request Models ────────────────────────────────────

class ProjectCreate(BaseModel):
    title: str
    description: Optional[str] = None
    total_budget: Optional[Decimal] = Field(None, ge=0, decimal_places=2)
    start_date: Optional[date] = None
    end_date: Optional[date] = None

class ProjectUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    total_budget: Optional[Decimal] = Field(None, ge=0, decimal_places=2)
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    status: Optional[ProjectStatus] = None


# ── Response Models ───────────────────────────────────

class ProjectResponse(BaseModel):
    id: str
    org_id: str
    title: str
    description: Optional[str] = None
    total_budget: Optional[Decimal] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    status: ProjectStatus
    progress_pct: Decimal = Decimal("0")
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime

class ProjectSummary(BaseModel):
    """Lightweight project summary for list views."""
    id: str
    title: str
    status: ProjectStatus
    progress_pct: Decimal
    total_budget: Optional[Decimal] = None
    worksite_count: int = 0
    start_date: Optional[date] = None
    end_date: Optional[date] = None

class ProjectDetail(ProjectResponse):
    """Full project detail with nested worksites."""
    worksites: list = []       # List[WorksiteResponse] — circular import handled at runtime
    worksite_count: int = 0
    total_invoiced: Decimal = Decimal("0")
    total_paid: Decimal = Decimal("0")

File: frontend/src/types/project.ts
export interface Project {
  id: string;
  org_id: string;
  title: string;
  description: string;
  location: string;
  budget: number;
  start_date: string;
  deadline: string;
  status: ProjectStatus;
  progress_pct: number;
  custom_fields?: Record<string, unknown>;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export type ProjectStatus =
  | 'draft'
  | 'planning'
  | 'active'
  | 'review'
  | 'complete'
  | 'on_hold'
  | 'cancelled';

export interface CreateProjectRequest {
  title: string;
  description: string;
  location: string;
  budget: number;
  start_date: string;
  deadline: string;
  custom_fields?: Record<string, unknown>;
}

export interface UpdateProjectRequest {
  title?: string;
  description?: string;
  location?: string;
  budget?: number;
  start_date?: string;
  deadline?: string;
  status?: ProjectStatus;
  custom_fields?: Record<string, unknown>;
}
